Fix Turf Transparency for MultiZ

timothymtorres opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

Fixes #62794 
Fixes #57654 
Fixes #53666

~~This might fix issue #61998~~ Nevermind.  It does not.

Transparent turfs, catwalks, and lighting overlays were badly broken with MultiZ issues.  This fixes any transparent turf to appear and be on the same z-level if it has the `turf_transparent` element.  Catwalks will appear and be on the same z-level as well.  They were initially glitching to top z-levels causing people to walk over an illusion that caused them to fall to their deaths.  

After I fixed these issues I noticed that lighting overlays were also glitching to top z-levels, which looked rather strange as there were lights in the middle of rooms and hallways with no supporting walls.  Lights now appear under the transparent turfs properly. 

The mapping icons for catwalks displayed the wrong `icon_state` so I fixed that too.

I made a big refactor for transparency and catwalks code that had to be done to get all of this to work properly.
## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

You no longer have to be afraid of falling off of glass floors, catwalks, or transparent tiles.  Also lighting overlays don't glitch through upper z-levels anymore for transparent tiles.

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
fix: Fix glass floors appearing as lower z-level when on the same z-level.
fix: Fix catwalks appearing as current z-level when they were on a lower z-level.
fix: Fix transparent turfs to display lower z-level properly.
fix: Fix lighting overlays appearing as current z-level when they were on lower z-level. (for transparent turfs)
fix: Fix mapping icons for catwalks to display correct icon state
refactor: Refactor turf transparency to be more robust.
refactor: Refactor catwalks to be elegant code.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

timothymtorres committed:
Fix glass floors not appearing on correct planet/layer

timothymtorres committed:
Remove catwalks on bottom z-level in underpass

Kylerace commented:
this doesnt fix https://github.com/tgstation/tgstation/issues/62794 since placing those turfs after mapload would still break things

timothymtorres commented:
> this doesnt fix https://github.com/tgstation/tgstation/issues/62794 since placing those turfs under openspace after mapload would still break things

I removed the catwalks on the  bottom z level and replaced them with the yellow square decals.  Glass floors I only changed the code.

Ghommie commented:
The open space turf (or its plane) should have a slight blur filter. It looks pretty flat right now.

timothymtorres commented:
> The open space turf (or its plane) should have a slight blur filter. It looks pretty flat right now.

That has nothing to do with my PR.

The PR that was going to add blur effects to open space never got merged bc it went stale.

MrMelbert commented:
The issue of #62794 involves catwalk PLATING - `/turf/open/floor/catwalk_floor`, not catwalk lattice - `/obj/structure/lattice/catwalk`

You didn't touch catwalk plating in this PR, so the issue will still be prevalent

timothymtorres commented:
> The issue of #62794 involves catwalk PLATING - `/turf/open/floor/catwalk_floor`, not catwalk lattice - `/obj/structure/lattice/catwalk`
> 
> 
> 
> You didn't touch catwalk plating in this PR, so the issue will still be prevalent 

The issue title is misleading.  The catwalks they are referencing are actually the structure type.  See the file changes for the tram map.  All the catwalks were removed.

Ghommie commented:
> 
> That has nothing to do with my PR.

Oh sorry, that was a general suggestion written in a hurry because I had to go somewhere. It's not specific to this PR.

> 
> The PR that was going to add blur effects to open space never got merged bc it went stale.

Good to know. I'll give it a look later.

LemonInTheDark commented on _maps/map_files/tramstation/tramstation.dmm:
```diff
@@ -6586,11 +6586,6 @@
 /obj/effect/decal/cleanable/dirt,
 /turf/open/floor/plating,
 /area/maintenance/tram/left)
-"aMM" = (
-/obj/structure/lattice/catwalk,
-/obj/effect/decal/cleanable/dirt,
-/turf/open/floor/plating,
-/area/maintenance/tram/left)
```
This is dumb. If they look wrong that's a failing of the system. rather then something to be hid

timothymtorres replied:
These changes were removed.

LemonInTheDark commented on code/game/turfs/open/glass.dm:
```diff
@@ -27,6 +27,8 @@
 /turf/open/floor/glass/LateInitialize()
 	. = ..()
 	AddElement(/datum/element/turf_z_transparency, TRUE)
+	plane = FLOOR_PLANE
+	layer = TURF_LAYER
```
Not changing the plane should be an option attached to the element, rather then being hardcoded like this

timothymtorres replied:
This has been changed.

LemonInTheDark commented:
> 
> The issue title is misleading. The catwalks they are referencing are actually the structure type. See the file changes for the tram map. All the catwalks from the issue were removed from the underpass section so players don't get confused.

No. They gave the structure type as an example of something that works, rather then the issue. They were discussing catwalk tiles, these things
![image](https://user-images.githubusercontent.com/58055496/142091228-8056e27c-8ca1-4ba6-89cd-1d6265923998.png)

timothymtorres committed:
Add is_openspace arg to turf_transparency element

timothymtorres committed:
Add defines for openspace transparency and z level

timothymtorres committed:
Remove resetting plane and layer vars

timothymtorres commented:
> > The issue title is misleading. The catwalks they are referencing are actually the structure type. See the file changes for the tram map. All the catwalks from the issue were removed from the underpass section so players don't get confused.
> 
> No. They gave the structure type as an example of something that works, rather then the issue. They were discussing catwalk floors, these things ![image](https://user-images.githubusercontent.com/58055496/142091228-8056e27c-8ca1-4ba6-89cd-1d6265923998.png) ![image](https://user-images.githubusercontent.com/58055496/142091295-96df36e6-9272-44e2-a2c7-e203fdcad3a3.png) Kinda unclear though, since catwalk floors isn't a very distinctive name

I tested these catwalks and they appear on the top z-level when they are in fact on the bottom z-level.  I'm in the process of refactoring how the multi-z code is written to iron out these problems.

timothymtorres committed:
Change catwalk icon_state for easier mapping

timothymtorres committed:
Remove defines from openspace element

timothymtorres committed:
Remove unused boolean transparency arg

timothymtorres committed:
Refactor transparency procs and vars to be sane

Ghommie commented on code/datums/elements/turf_transparency.dm:
```diff
@@ -1,29 +1,33 @@
+		our_turf.vis_contents.len = 0
+		if(!add_baseturf_underlay(our_turf))
+			our_turf.ChangeTurf(/turf/open/floor/plating, flags = CHANGETURF_INHERIT_AIR)
+			return FALSE
```
```suggestion
			return ELEMENT_INCOMPATIBLE
```
Attach shouldn't have binary return values iirc.

timothymtorres replied:
This was resolved.

timothymtorres committed:
Refactor transparency to be less verbose

timothymtorres committed:
Remove transparency args

timothymtorres committed:
Merge branch 'master' into tramstation-catwalk-hotfix

timothymtorres committed:
Revert tram changes

timothymtorres committed:
Fix lighting overlays appearing on top multi-z turfs

timothymtorres committed:
Fix catwalks multi-z problems

timothymtorres committed:
Add plating underlay to catwalk and fix mapping icons

timothymtorres committed:
Change icon_state to use catwalk_type

timothymtorres committed:
Revert all tramstation testing changes

timothymtorres committed:
Merge branch 'master' into tramstation-catwalk-hotfix

timothymtorres committed:
Revert ANY CHANGES to tramstation

Come on, come on, work this time!

timothymtorres committed:
REVERT THE CHANGES GOD DAMN GIT

timothymtorres committed:
WHY WONT YOU REVERT

timothymtorres committed:
ONE MORE TRY

timothymtorres committed:
will the real tramstation please stand up

timothymtorres committed:
Delete tramstation.dmm

timothymtorres committed:
please dear god just revert

timothymtorres commented:
@LemonInTheDark it's finally ready and the original post has been updated with info.  I tested it pretty thoroughly on local with no issues that I could find.

LemonInTheDark commented on code/datums/elements/turf_transparency.dm:
```diff
@@ -1,27 +1,28 @@
-
-	our_turf.plane = OPENSPACE_PLANE
 	our_turf.layer = OPENSPACE_LAYER
+	if(istype(our_turf, /turf/open/openspace))
+		our_turf.plane = OPENSPACE_PLANE
```
Do not like this. Why nuke the argument?

timothymtorres replied:
Several important reasons:

`show_bottom_level` was used both as a argument AND as a proc with the same name. This was a mindfuck and was creating confusion as to what is being used where.  Also the name was misleading for both cases.

So then why didn't I just rename the argument to something else?  Well I did initially, but then it turns out the only time this arg was actually getting used was to check if a turf was `/turf/open/openspace` since I don't ever foresee a new turf that behaves like openspace anytime soon, we don't need to make it into an arg.  Also to back that claim up, it's been 5 years since this code was made and there still isn't any new types of openspace.  It'd be like someone making a different version of `/turf/open/space`.  So hardcoding the check in this case is fine and the arg only gets used in two spots.

LemonInTheDark replied:
I meant is_openspace, which while it should maybe be renamed seemed valuable.
Being able to control if a tile should join the openspace plane or not is something I forsee being helpful.
The same can actually be said of show_bottom_level, unfucking multiz space is on my bucket list

timothymtorres replied:
The `openspace` turf is one of a kind.  It's similar to bedrock blocks for minecraft.  We do not want stuff falling through a z-level that is supposed to be rock bottom.  Make sense?  That's what the typecheck is for.  No other turf is going to inherit this behaviour that I can think of now or ever.  Thus it behaves like bedrock.  If anything, maybe a subclass might be made (I don't see why) and this check would still be valid to work with the new subclass.  

Keep in mind, it's been 5 years this code has been here and nothing has changed.  There is no point in making this a modular piece of code via making it into an argument or a trait.

We do however want multi Z transparency to be modular. (which is what this PR helps to do)

LemonInTheDark replied:
I don't see any reason it couldn't be supported. Depending on how profiles work out it's something I'm considering using with #63255, and I'd like to have that option open.

Is it just a cleanliness thing? or is there a deeper issue behind it

LemonInTheDark commented on code/datums/elements/turf_transparency.dm:
```diff
@@ -56,20 +53,20 @@
-	if(!show_bottom_level)
-		return FALSE
+/datum/element/turf_z_transparency/proc/add_baseturf_underlay(turf/our_turf)
+	if(istype(our_turf, /turf/open/openspace)) // we don't ever want our bottom turf to be openspace
```
Also do not like this, but I think I get what you're going for

LemonInTheDark replied:
We don't always want to reset to base right? or was the old intent of that check wrong

timothymtorres replied:
I had to rename the proc because it was misleading and conflicting with another variable that had the same name.

If the bottom z-level is openspace then a mapper made an opsie.  This fixes any openspace tiles on the bottom z-level to be changed to plating as a safety precaution.  I should probably throw in some logging since if this is happening either an admin or a mapper did something wrong since it shouldn't be happening.

LemonInTheDark replied:
I meant the typecheck, it seems fragile to assume that typepath is the only unacceptable bottom

timothymtorres replied:
See my above comment about openspace behaviour resembling bedrock in minecraft.

LemonInTheDark commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -7,37 +7,30 @@
 
 /turf/open/floor/catwalk_floor/Initialize(mapload)
 	. = ..()
+	var/mutable_appearance/plating_underlay = mutable_appearance(icon, "[catwalk_type]_below", TURF_LAYER)
+	underlays += plating_underlay
```
this is a bit sus, I'd think the same caching issue applies here

LemonInTheDark replied:
Why underlays btb?

timothymtorres replied:
I ditched the fancy caching since performance _shouldn't_ be critical for this code.  

For something like atmos code it makes sense to optimize it where we can.  But for catwalk initialization you might squeeze a fraction of a second during bootup times with the drawback making the code really ugly.

```
GLOBAL_LIST_EMPTY(catwalk_overlay_masterlist) //Stores all the above_states for the different types of catwalk

if(!GLOB.catwalk_overlay_masterlist[above_state])
		GLOB.catwalk_overlay_masterlist[above_state] = catwalk_overlay
	. += GLOB.catwalk_overlay_masterlist[above_state]
```

This is all really ugly and unnecessary just to give us a meager boost in performance where it doesn't matter. 

Also I suspect the reason caching was used before was that the overlay was being constantly created and deleted for each catwalk individually.  The masterlist created one overlay to be used for all catwalks.  

With my code refactor, catwalks have an underlay that is never deleted.  Only the icon_state is changed.  Technically, I could cache the underlays during init, but as I said before I think trading code cleanliness for performance is bad design.  (exceptions are performance bottlenecks)

LemonInTheDark replied:
we cache overlays like that not just to save on cpu cycles, but also to save on memory (Though to be straight up with you I do not remember if overlays benefit from that in the same way as vis_contents)

timothymtorres replied:
I posted in the comment sections the memory footprint for before and after.  The result was negligible.

LemonInTheDark replied:
It's about the principle of the thing, unless there's a functional reason not to

timothymtorres replied:
> When I was doing my testing, I noticed that the layers and planes would break only for overlays/underlays/mutable_appearances. Thus we only want to apply those updates to the main object directly. The way the code was before, the catwalk overlay was updating the plane/layer causing it to break. I reversed it so that the underlay is the bottom portion of the catwalk and the regular object is the upper portion with the grilles that switch on/off.


Quoted from above,

When I removed the `!GLOB.catwalk_overlay_masterlist` I did try to keep the plating as the main icon and update the overlay.  It wouldn't work.  So had to reverse their roles and make the main icon the catwalk grille and the underlay the plating.  Updating the layers/planes for overlays breaks the Multi Z when I tested this about a month ago.  

`Icons are for image manipulation`

All the manipulation is being performed on the icon, which is the catwalk plating top portion.  The bottom plating part never changes and is an underlay.

LemonInTheDark replied:
Generating a new appearance for each catwalk when we only really ever need one copy seems silly, and I'm concerned about setting a good example on how to work with the things. I'm frankly not too sure if it has an impact, but I know if the pattern were continued elsewhere we'd have a headache, so I'd like to avoid it

timothymtorres replied:
Let me know if https://github.com/timothymtorres/tgstation/commit/c16ba26e27fd0ac6d57f52afc0bb3c82551046b7 is satisfactory.

LemonInTheDark commented on code/game/turfs/open/floor/catwalk_plating.dm:
```diff
@@ -7,37 +7,30 @@
 /turf/open/floor/catwalk_floor/update_overlays()
+	layer = covered ? CATWALK_LAYER : TURF_LAYER
+	plane = covered ? GAME_PLANE : FLOOR_PLANE
+	icon_state = covered ? "[catwalk_type]_above" : "[catwalk_type]_below"
```
Why are you updating the icon state in update overlays?

timothymtorres replied:
I wasn't aware there was a different proc to do this in.  The updates were happening here before so I kept the location the same.

LemonInTheDark replied:
You should look at how update_appearance works. There's different procs for different types of updates, and you can request one of the other depending on the situation

LemonInTheDark replied:
On another note, what was the actual issue here? the use of overlays for some reason?

timothymtorres replied:
> You should look at how update_appearance works.

Sounds good and I will try this.

> On another note, what was the actual issue here? the use of overlays for some reason?

Overlays for lighting and catwalks were glitching through the layers/planes to the current z-level the player was in.  When `transparent_turf` element was attached to a turf, the layer would be set to the `OPENSPACE_LAYER` which resulted in it being drawn as a bottom z-level with shadows around it.  This was more noticeable with glass floors.

NanoCats commented:
> > The issue title is misleading. The catwalks they are referencing are actually the structure type. See the file changes for the tram map. All the catwalks from the issue were removed from the underpass section so players don't get confused.
> 
> No. They gave the structure type as an example of something that works, rather then the issue. They were discussing catwalk floors, these things Kinda unclear though, since catwalk floors isn't a very distinctive name



Terribly sorry for the confusion, but please note that these things might deserve different names, if someone says "Catwalk" they could conceivably be referring to:

- Catwalk (Latice) 
- Catwalk (Catwalk) (NOTE THAT AS OF A RECENT PR YOU CAN APPARENTLY RETEXTURE CATWALK TILES TO LOOK LIKE THESE TOO)
- Catwalk (Catwalk _tile_)
- Potential fourth thing also called a catwalk

LemonInTheDark commented:
Undrafting when tim gets home from vacation (stop using github while you're on break tim)

timothymtorres committed:
Change catwalks to use update_appearance()

timothymtorres commented:
@LemonInTheDark this is the memory footprint (on tram) before the changes (with caching)

![dreamdaemon_7tArzt5XsS](https://user-images.githubusercontent.com/5195984/146370776-e9b5d72e-c0e9-4c54-8a29-36e28dd76786.png)

This is the memory footprint after my changes (with caching removed)

![dreamdaemon_ZYHcDA7nqn](https://user-images.githubusercontent.com/5195984/146370917-b7845548-7e3b-41a3-b727-2bfb2f1ac067.png)

timothymtorres commented on code/datums/elements/turf_transparency.dm:
```diff
@@ -31,15 +32,11 @@
 	REMOVE_TRAIT(our_turf, TURF_Z_TRANSPARENT_TRAIT, ELEMENT_TRAIT(type))
 
 ///Updates the viscontents or underlays below this tile.
-/datum/element/turf_z_transparency/proc/update_multiz(turf/our_turf, prune_on_fail = FALSE, init = FALSE)
```
@LemonInTheDark 

> I don't see any reason it couldn't be supported. Depending on how profiles work out it's something I'm considering using with #63255, and I'd like to have that option open.
> 
> Is it just a cleanliness thing? or is there a deeper issue behind it

I gutted these args because all it was doing was checking to see if the turf was openspace.  I moved these checks to `istype()` procs since IMO it was more obvious why the check was happening than passing it as an arg.  Mainly it was cleanup to improve readability.  

Since `/turf/open/openspace/space` is a subclass of `/turf/open/openspace` there shouldn't be any problems.  If you are really deadset in making this modular I can add the args back, I just don't see any benefit.

LemonInTheDark replied:
typechecks like that smell a bit to me, specially for an element like this. IDK, it's not like other turfs couldn't in theory want that behavior, they just don't right now. 

Don't want people repeating what happen with shoving, or just sticking another typecheck in and moving on. Prefer things as clear as possible.

Maybe the vars need better documentation as to their purpose? idk

timothymtorres replied:
Fixed with https://github.com/tgstation/tgstation/pull/62875/commits/31a98491f97acd8bc04156741821ed6e95d1153e

https://github.com/tgstation/tgstation/blob/9f9f633ebd322596cccfd05ab19793ce03e46af3/code/__DEFINES/is_helpers.dm#L15-L23

The other turfs used by the helper won't ever have this element attached to it. (space, chasm, lava, water)  If this is a problem then I can make a seperate is_helper proc just for openspace turfs.  This change should also be compatible with https://github.com/tgstation/tgstation/pull/63559 as long as they add their new turf type to the is_helper list.

And just FYI, this change is untested (I'm webediting) but small enough that it shouldn't need it.

timothymtorres committed:
Change multiZ transparency to use isopenturf proc

timothymtorres committed:
Fix multi-Z transparency to use isgroundless proc

timothymtorres commented:
The typecast has been fixed now and should include the other space/openspace type by using a is helper.

LemonInTheDark commented on code/datums/elements/turf_transparency.dm:
```diff
@@ -56,20 +53,20 @@
-	if(!show_bottom_level)
-		return FALSE
+/datum/element/turf_z_transparency/proc/add_baseturf_underlay(turf/our_turf)
+	if(isgroundlessturf(our_turf)) // we don't ever want our bottom turf to be openspace
+		our_turf.ChangeTurf(/turf/open/floor/plating, flags = CHANGETURF_INHERIT_AIR)
```
I really am not a fan of this, and would prefer it to just be a var option on the element add, unless there's a strong reason to not do that.

Typechecks in component initializes is fine, we often do it so a component can support some general "idea", with unique implementations for different types of objects. Only inhand items can honk on use for instance.

I don't like this sort of pattern, because it's tying the behavior of the element to something that's not inherent to the types you're checking for, and may even play poorly with them.

Lava for instance may not want to be scraped to a plating, or want to receive the openspace plane. Going from this to a bespoke typecache for multiz transparency only also seems fraught, since then rather then changing the arguments of the element, we're forced to seek out the correct list of types and add ourselves to it. Makes things easier to fuck up.

I realize I'm being hardheaded about this, but it seems like the wrong way to go with this solution. I realize the clutter you were trying to solve, but while removing useless args is very good, I'm not really convinced this is in the same category.

timothymtorres replied:
Alright I'll look into fixing this tonight

timothymtorres replied:
Alright it's been updated with the proposed changes.

LemonInTheDark reviewed (changes requested):
Once more, hopefully for the last time, sorry for taking so long/being such a hardhead about this

timothymtorres committed:
Add static list for catwalk underlays

timothymtorres committed:
Change turf_z_transparency to use is_openspace arg

timothymtorres committed:
Add is_openspace arg to transparency element
